[pytree] add key path api

suo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #116786

This PR introduces a key path API to pytrees, drawing direct inspiration from JAX's [key path API](https://jax.readthedocs.io/en/latest/jax-101/05.1-pytrees.html#key-paths).

I added the 3 APIs described there, and a registry of `flatten_with_keys` fns for each node type, which is a version of `flatten` that also returns `KeyEntry`s describing how to access values from the original pytree.

Current use cases for this API:
- Folks would like to do argument traversal over input pytrees to do verification and compatibility enforcement. Keypaths are useful for this—https://fburl.com/code/06p7zrvr is a handrolled pass doing basically the same thing but probably more fragilely.
- In export non-strict mode, we need to figure out a way to track sources for pytree inputs. In strict mode, dynamo handles this for us, but we'd like a decoupled component to handle this when we're not using dynamo.

I'm sure there are places it would be useful.

Some design notes:
- I only implemented the API for  the Python pytree impl. optree has some differences in how their keypath APIs are designed (see https://github.com/pytorch/pytorch/issues/113378 for discussion). I have some issues with the proposed typed_path solution in that discussion and prefer JAX's API, but we can hash that out separately.
- The way folks register a `flatten_with_keys` fn is through a new kwarg to `register_pytree_node`. This follows how we do serialization fns, although the list of additional arguments is getting unwieldy.
- My impl handles pytrees with an undefined `flatten_with_keys` fn is different from JAX. I will raise an error, JAX creates a fallback keyentry.

Differential Revision: [D52547850](https://our.internmc.facebook.com/intern/diff/D52547850/)

suo committed:
[pytree] add key path api

This PR introduces a key path API to pytrees, drawing direct inspiration from JAX's [key path API](https://jax.readthedocs.io/en/latest/jax-101/05.1-pytrees.html#key-paths).

I added the 3 APIs described there, and a registry of `flatten_with_keys` fns for each node type, which is a version of `flatten` that also returns `KeyEntry`s describing how to access values from the original pytree.

Current use cases for this API:
- Folks would like to do argument traversal over input pytrees to do verification and compatibility enforcement. Keypaths are useful for this—https://fburl.com/code/06p7zrvr is a handrolled pass doing basically the same thing but probably more fragilely.
- In export non-strict mode, we need to figure out a way to track sources for pytree inputs. In strict mode, dynamo handles this for us, but we'd like a decoupled component to handle this when we're not using dynamo.

I'm sure there are places it would be useful.

Some design notes:
- I only implemented the API for  the Python pytree impl. optree has some differences in how their keypath APIs are designed (see https://github.com/pytorch/pytorch/issues/113378 for discussion). I have some issues with the proposed typed_path solution in that discussion and prefer JAX's API, but we can hash that out separately.
- The way folks register a `flatten_with_keys` fn is through a new kwarg to `register_pytree_node`. This follows how we do serialization fns, although the list of additional arguments is getting unwieldy.
- My impl handles pytrees with an undefined `flatten_with_keys` fn is different from JAX. I will raise an error, JAX creates a fallback keyentry.

Differential Revision: [D52547850](https://our.internmc.facebook.com/intern/diff/D52547850/)

[ghstack-poisoned]

voznesenskym commented on torch/utils/_pytree.py:
```diff
@@ -282,11 +309,44 @@ def _private_register_pytree_node(
+
+
+@dataclasses.dataclass(frozen=True)
+class SequenceKey:
```
nit/bikeshed: IndexKey (more correct) or GetItemKey (matches GetAttrKey) or ListIndex/Key (Matches DictKey)

suo replied:
I changed this to map to the collections.abc scheme (SequenceKey, MappingKey). GetAttrKey keeps its name because there is no abstract base class that represents "bag of attributes" and I couldn't think of a better name.

voznesenskym commented on torch/utils/_pytree.py:
```diff
@@ -89,6 +90,9 @@
 FromDumpableContextFn = Callable[[DumpableContext], Context]
 ToStrFunc = Callable[["TreeSpec", List[str]], str]
 MaybeFromStrFunc = Callable[[str], Optional[Tuple[Any, Context, str]]]
+KeyEntry = Any  # Should be Hashable
+KeyPath = Tuple[KeyEntry, ...]
```
Can this be something more concrete? Since the original discussion was on sources, 2 quesitons:

1) Why not use sources directly?

2) And if rejecting sources directly, should we have a source like object in that it has both the base and the accessor together? So instead of DictKey being something like `DictKey("foo")`, it would be `DictKey(base="my_dict", key="foo")`

suo replied:
> Why not use sources directly?

Mostly a layering thing; I consider pytree to be a lower level component that shouldn't necessarily "know" about higher-level concepts like shape env/dynamo/sources. Ideally these concepts map onto each other fairly closely.

> And if rejecting sources directly, should we have a source like object in that it has both the base and the accessor together? So instead of DictKey being something like DictKey("foo"), it would be DictKey(base="my_dict", key="foo")

Great question, let me think about that. I rather uncritically replicated the JAX API but maybe it can be improved on for our use case.

voznesenskym replied:
Sources are a lot lower level than dynamo or shape env imo - they are a primitive that fulfills the invariant of `eval(source.name())` giving you back the user object referred to in the source. 

We should reuse them where possible, especially as we start chatting about the generic pt2 agnostic guards/specialization system (as we have).

zou3519 replied:
With Voz's original comment over chat, I thought that we were going to write a utility to track sources through pytree.tree_flatten. The disadvantage of the key path API is that we now need users to provide key path rules for their custom pytree types.

To avoid the layering issue: pytree.tree_flatten shouldn't use sources in its implementation. Would it be possible to do something like turn on Dynamo's source tracking and call pytree.tree_flatten to obtain the sources?

suo replied:
That seems like an even worse layering violation haha; a system that wants only to use pytrees must now opt into using dynamo in order to use a relatively simple functionality. 

I will investigate the sources stuff to see if it can be pulled out of dynamo.

zou3519 replied:
Independently I think the key-path API is useful by itself, but it seems more robust for the AOTExport use case to not rely on users defining key path rules for custom pytree classes

suo replied:
Yeah it's an extra step, but I think it's okay. Users need to specify serialization rules as well anyway. 

If we are really worried about this, is to implement the flat container fallback that JAX has. If a `flatten_with_keys` fn is not defined, then we when we try to flatten with paths, a special `KeyEntry` is returned that signals that we first need to call `tree_flatten` and then index into the flattened list to retrieve the value. This trades efficiency for guaranteed coverage.

In general, I prefer to be explicit vs. implicit with the export APIs, so I didn't implement it that way here, but I think it's a reasonable change to make since key paths are not export-specific.

As a related note, JAX has a nice class-based registration API that I like that we should consider implementing. That is a slightly more robust way to do registration, since it does not separate registration from class definition (and thus increase the likelihood that the class is changed but not the pytree registration functions).

voznesenskym replied:
Discussed offline/chat. We agreed that if we do this, we want generic sources free of dynamo notions. As such, we would either move the guard_source,reconstruct stuff to a helper fn that visits the source types and dispatches behavior based on them OR as a mixin to provide the dynamo functions to implement, but allowing for other systems to not know about the dynamo side of the type.

ezyang replied:
@voznesenskym bat signaled me, so here is my spicy take.

I think that strict / non-strict should have the same way of describing the structure on input types. However, in both cases, I think we should *not* use Dynamo sources. Dynamo sources work on arbitrary Python data structure, not only those that have pytree rules defined for them, and they are abstraction breaking (e.g., if a tensor is stored on a private field `_foo`, we will just slap together a source `x._foo` in this case.) In contrast, export input description is part of the public API, so you want it simple, and you need some abstraction mechanisms that allow custom data types to say how exactly they would like to present their internal structure (and not just blat out all their guts.)

I think it's probably OK for export to use keypaths. It's pretty simple and there's already a custom registration system you can lift off. But then strict mode should still use keypaths too!

voznesenskym commented on torch/utils/_pytree.py:
```diff
@@ -282,11 +309,44 @@ def _private_register_pytree_node(
+
+
+@dataclasses.dataclass(frozen=True)
+class GetAttrKey:
```
If this (and the above) are KeyEntry, we should make that typing explicit.

suo replied:
I changed KeyEntry to be a hashable protocol, so we're using structural subtyping for this now

suo commented:
Hm okay, in the interest of moving things forward, here's a concrete proposal:
1. We can start by using a separate `KeyEntry` abstraction for pytrees.
1. We can do the work to migrate torch.export (in strict and non-strict forms) to use keypaths, so that the input handling is the same for all export modes.
1. We can keep `Source` as is for now, and produce some mapping of keypaths onto `Source`s so that things work with Dynamo. We can decide if we want to canonicalize an explicit mapping, or trace through keypath access with dynamo, or whatever.

Reading @ezyang's take I think I agree that KeyPaths and Sources aren't the same thing. They both represent access to a structure, but the usage is quite different: KeyPaths are explicit (user-defined) and Sources are implicit (just faithfully record what's going on in Python). 

At some point, we could develop a general-purpose accessor abstraction that can be re-used either in Dynamo (and enriched there with codegen concepts, etc.) and in pytree. But I'm not quite sure what that abstraction would be buying us at that point; I'm struggling to imagine a situation where you could write code using that abstraction where you didn't have to think about whether you're hitting dynamo in the end.

suo committed:
Update on "[pytree] add key path api"


This PR introduces a key path API to pytrees, drawing direct inspiration from JAX's [key path API](https://jax.readthedocs.io/en/latest/jax-101/05.1-pytrees.html#key-paths).

I added the 3 APIs described there, and a registry of `flatten_with_keys` fns for each node type, which is a version of `flatten` that also returns `KeyEntry`s describing how to access values from the original pytree.

Current use cases for this API:
- Folks would like to do argument traversal over input pytrees to do verification and compatibility enforcement. Keypaths are useful for this—https://fburl.com/code/06p7zrvr is a handrolled pass doing basically the same thing but probably more fragilely.
- In export non-strict mode, we need to figure out a way to track sources for pytree inputs. In strict mode, dynamo handles this for us, but we'd like a decoupled component to handle this when we're not using dynamo.

I'm sure there are places it would be useful.

Some design notes:
- I only implemented the API for  the Python pytree impl. optree has some differences in how their keypath APIs are designed (see https://github.com/pytorch/pytorch/issues/113378 for discussion). I have some issues with the proposed typed_path solution in that discussion and prefer JAX's API, but we can hash that out separately.
- The way folks register a `flatten_with_keys` fn is through a new kwarg to `register_pytree_node`. This follows how we do serialization fns, although the list of additional arguments is getting unwieldy.
- My impl handles pytrees with an undefined `flatten_with_keys` fn is different from JAX. I will raise an error, JAX creates a fallback keyentry.

Differential Revision: [D52547850](https://our.internmc.facebook.com/intern/diff/D52547850/)

[ghstack-poisoned]

suo committed:
Update on "[pytree] add key path api"


This PR introduces a key path API to pytrees, drawing direct inspiration from JAX's [key path API](https://jax.readthedocs.io/en/latest/jax-101/05.1-pytrees.html#key-paths).

I added the 3 APIs described there, and a registry of `flatten_with_keys` fns for each node type, which is a version of `flatten` that also returns `KeyEntry`s describing how to access values from the original pytree.

Current use cases for this API:
- Folks would like to do argument traversal over input pytrees to do verification and compatibility enforcement. Keypaths are useful for this—https://fburl.com/code/06p7zrvr is a handrolled pass doing basically the same thing but probably more fragilely.
- In export non-strict mode, we need to figure out a way to track sources for pytree inputs. In strict mode, dynamo handles this for us, but we'd like a decoupled component to handle this when we're not using dynamo.

I'm sure there are places it would be useful.

Some design notes:
- I only implemented the API for  the Python pytree impl. optree has some differences in how their keypath APIs are designed (see https://github.com/pytorch/pytorch/issues/113378 for discussion). I have some issues with the proposed typed_path solution in that discussion and prefer JAX's API, but we can hash that out separately.
- The way folks register a `flatten_with_keys` fn is through a new kwarg to `register_pytree_node`. This follows how we do serialization fns, although the list of additional arguments is getting unwieldy.
- My impl handles pytrees with an undefined `flatten_with_keys` fn is different from JAX. I will raise an error, JAX creates a fallback keyentry.

Differential Revision: [D52547850](https://our.internmc.facebook.com/intern/diff/D52547850/)

[ghstack-poisoned]

voznesenskym commented:
> Hm okay, in the interest of moving things forward, here's a concrete proposal:
> 
> 1. We can start by using a separate `KeyEntry` abstraction for pytrees.
> 2. We can do the work to migrate torch.export (in strict and non-strict forms) to use keypaths, so that the input handling is the same for all export modes.
> 3. We can keep `Source` as is for now, and produce some mapping of keypaths onto `Source`s so that things work with Dynamo. We can decide if we want to canonicalize an explicit mapping, or trace through keypath access with dynamo, or whatever.
> 
> Reading @ezyang's take I think I agree that KeyPaths and Sources aren't the same thing. They both represent access to a structure, but the usage is quite different: KeyPaths are explicit (user-defined) and Sources are implicit (just faithfully record what's going on in Python).
> 
> At some point, we could develop a general-purpose accessor abstraction that can be re-used either in Dynamo (and enriched there with codegen concepts, etc.) and in pytree. But I'm not quite sure what that abstraction would be buying us at that point; I'm struggling to imagine a situation where you could write code using that abstraction where you didn't have to think about whether you're hitting dynamo in the end.

this SGTM - I wasn't so much trying to shove sources as have a good enough reason to not use them - this feels like it's more than good enough. 

For the KeyEntry objects, question 2 above is still worth answering

voznesenskym reviewed (approved):
This LGTM - posting the question that I had above

> should we have a source like object in that it has both the base and the accessor together? So instead of DictKey being something like DictKey("foo"), it would be DictKey(base="my_dict", key="foo")

And @suo's response 

> Great question, let me think about that. I rather uncritically replicated the JAX API but maybe it can be improved on for our use case.

This is more of a stylistic thing, I dont have strong feelings about it looking either which way.

suo commented:
Oops I missed that one!

> And if rejecting sources directly, should we have a source like object in that it has both the base and the accessor together? So instead of DictKey being something like DictKey("foo"), it would be DictKey(base="my_dict", key="foo")

I think the API difference is that KeyPath is the primary abstraction instead of KeyEntry (which is the object analogous to Source). A KeyPath is a sequence that describes accesses, which naturally chains them together already. It looks like:
```
(SequenceKey(5), MappingKey("lol")) -> obj[5]["lol"]
```

All the APIs take/return KeyPaths, so if you add a method to KeyEntry like:
```
    def access(self, parent: Any) -> Any:
        ...
```

then chained access can be implemented as simply as: 
```
def key_get(obj: Any, kp: KeyPath) -> Any:
    for k in kp:
        obj = k.access(obj)
    return obj
```

I have a followup PR that implements this.

XuehaiPan commented on torch/utils/_pytree.py:
```diff
@@ -1181,3 +1300,91 @@ def arg_tree_leaves(*args: PyTree, **kwargs: PyTree) -> List[Any]:
+    return list(_generate_key_paths((), tree))
+
+
+def _generate_key_paths(
+    key_path: KeyPath,
+    tree: Any,
+) -> Iterable[Tuple[KeyPath, Any]]:
```
Need to support extra custom leaf predicate function `is_leaf`. See also:

- #116419

```suggestion
def _generate_key_paths(
    key_path: KeyPath,
    tree: Any,
    is_leaf: Optional[Callable[[PyTree], bool]] = None
) -> Iterable[Tuple[KeyPath, Any]]:
```

suo replied:
awesome, didn't realize we added `is_leaf` to pytree!

XuehaiPan commented on torch/utils/_pytree.py:
```diff
@@ -1181,3 +1300,91 @@ def arg_tree_leaves(*args: PyTree, **kwargs: PyTree) -> List[Any]:
+        second element is a :class:`TreeSpec` representing the structure of the flattened
+        tree.
+    """
+    _, treespec = tree_flatten(tree)
+    return list(_generate_key_paths((), tree)), treespec
```
```suggestion
    treespec = tree_structure(tree)
    return list(_generate_key_paths((), tree)), treespec
```

I'm worried about the consistency between `flatten_fn` (used in `tree_{flatten,map}`) and `flatten_with_keys_fn` (used in `tree_{flatten,map}_with_path`). We need to make sure both functions produce the same `leaves` and `context`.

Another concern is performance, we firstly traverse the whole tree with `flatten_fn` first to get the `treespec`. Then we traverse the whole tree again with `flatten_with_keys_fn` again to get the paths. We can merge the two passes into one.

suo replied:
for consistency: I will add the test as mentioned above to check all the types we are registering as part of pytorch for consistency.

For performance: I think this could definitely be improved, but let's leave it for another PR I think. I think it's best to keep PR focused on the basic API design and functional correctness of the implementation, and I think modifying `_generate_key_paths` to also generate a treespec may make that harder.

suo commented on torch/utils/_pytree.py:
```diff
@@ -282,11 +323,44 @@ def _private_register_pytree_node(
+def _dict_flatten_with_keys(
+    d: Dict[Any, Any]
+) -> Tuple[List[Tuple[KeyEntry, Any]], Context]:
+    return list(zip(tuple(MappingKey(k) for k in d.keys()), d.values())), d.keys()
```
great point. I will add a test in `test_pytree.py` to check this.

suo committed:
Update on "[pytree] add key path api"


This PR introduces a key path API to pytrees, drawing direct inspiration from JAX's [key path API](https://jax.readthedocs.io/en/latest/jax-101/05.1-pytrees.html#key-paths).

I added the 3 APIs described there, and a registry of `flatten_with_keys` fns for each node type, which is a version of `flatten` that also returns `KeyEntry`s describing how to access values from the original pytree.

Current use cases for this API:
- Folks would like to do argument traversal over input pytrees to do verification and compatibility enforcement. Keypaths are useful for this—https://fburl.com/code/06p7zrvr is a handrolled pass doing basically the same thing but probably more fragilely.
- In export non-strict mode, we need to figure out a way to track sources for pytree inputs. In strict mode, dynamo handles this for us, but we'd like a decoupled component to handle this when we're not using dynamo.

I'm sure there are places it would be useful.

Some design notes:
- I only implemented the API for  the Python pytree impl. optree has some differences in how their keypath APIs are designed (see https://github.com/pytorch/pytorch/issues/113378 for discussion). I have some issues with the proposed typed_path solution in that discussion and prefer JAX's API, but we can hash that out separately.
- The way folks register a `flatten_with_keys` fn is through a new kwarg to `register_pytree_node`. This follows how we do serialization fns, although the list of additional arguments is getting unwieldy.
- My impl handles pytrees with an undefined `flatten_with_keys` fn is different from JAX. I will raise an error, JAX creates a fallback keyentry.

Differential Revision: [D52547850](https://our.internmc.facebook.com/intern/diff/D52547850/)

[ghstack-poisoned]

suo committed:
Update on "[pytree] add key path api"


This PR introduces a key path API to pytrees, drawing direct inspiration from JAX's [key path API](https://jax.readthedocs.io/en/latest/jax-101/05.1-pytrees.html#key-paths).

I added the 3 APIs described there, and a registry of `flatten_with_keys` fns for each node type, which is a version of `flatten` that also returns `KeyEntry`s describing how to access values from the original pytree.

Current use cases for this API:
- Folks would like to do argument traversal over input pytrees to do verification and compatibility enforcement. Keypaths are useful for this—https://fburl.com/code/06p7zrvr is a handrolled pass doing basically the same thing but probably more fragilely.
- In export non-strict mode, we need to figure out a way to track sources for pytree inputs. In strict mode, dynamo handles this for us, but we'd like a decoupled component to handle this when we're not using dynamo.

I'm sure there are places it would be useful.

Some design notes:
- I only implemented the API for  the Python pytree impl. optree has some differences in how their keypath APIs are designed (see https://github.com/pytorch/pytorch/issues/113378 for discussion). I have some issues with the proposed typed_path solution in that discussion and prefer JAX's API, but we can hash that out separately.
- The way folks register a `flatten_with_keys` fn is through a new kwarg to `register_pytree_node`. This follows how we do serialization fns, although the list of additional arguments is getting unwieldy.
- My impl handles pytrees with an undefined `flatten_with_keys` fn is different from JAX. I will raise an error, JAX creates a fallback keyentry.

Differential Revision: [D52547850](https://our.internmc.facebook.com/intern/diff/D52547850/)

[ghstack-poisoned]

zou3519 commented on torch/utils/_pytree.py:
```diff
@@ -400,18 +512,21 @@ def _deque_unflatten(values: Iterable[Any], context: Context) -> Deque[Any]:
     _tuple_flatten,
     _tuple_unflatten,
     serialized_type_name="builtins.tuple",
+    flatten_with_keys_fn=_tuple_flatten_with_keys,
```
The JAX API for registering pytrees with keypath API is different. There's a register_pytree_node and a separate register_pytree_node_with_keys. The reasoning behind this is that the flatten_with_keys_fn can be used in place of the flatten_fn. Is there a reason why we chose to diverge?

suo replied:
I diverged because we already have additional registration functions on `register_pytree_node` (the serialization stuff). Also, a separate registration call seems like it would make the API will be less discoverable, so people might forget to define a `flatten_with_keys_fn`—an issue that I have been bit with by with the `tree_flatten_spec` APIs which also have a separate registration API that no one uses lol.

The ability to re-use flatten_with_keys to implement flatten is nice though, so happy to change if you think it's worth it.

zou3519 commented on torch/utils/_pytree.py:
```diff
@@ -1250,3 +1369,120 @@ def arg_tree_leaves(*args: PyTree, **kwargs: PyTree) -> List[Any]:
+
+
+def keystr(kp: KeyPath) -> str:
+    """Give a key path, return a pretty-printed representation."""
```
nit: Given

zou3519 commented on torch/utils/_cxx_pytree.py:
```diff
@@ -825,3 +838,82 @@ def __instancecheck__(self, instance: object) -> bool:
+    Returns:
+        A list of (key path, leaf) pairs.
+    """
+    raise NotImplementedError("KeyPaths are not yet supported in cxx_pytree.")
```
Do we have a plan for adding these? @XuehaiPan, are you OK with how the keypath API looks like? There was some discussion over at https://github.com/pytorch/pytorch/issues/113378 about performance, but IMO we shouldn't diverge from JAX unless we have a good reason to.

suo replied:
happy to work with @XuehaiPan on C++ implementation. Agree that we should stay close to the JAX API.

XuehaiPan replied:
> Do we have a plan for adding these?
> happy to work with @XuehaiPan on C++ implementation

Yes, we should keep the Python and C++ pytrees to provide the same APIs. The current tree path implementation does not rely on the implementation detail of `tree_flatten` and `treespec`. You can just copy (or import) the key path API in Python pytree into C++ pytree and replace the `tree_flatten` with the C++ `tree_flatten`.

------

> @XuehaiPan, are you OK with how the keypath API looks like? There was some discussion over at #113378 about performance, but IMO we shouldn't diverge from JAX unless we have a good reason to.

> I agree that we should stay close to the JAX API.

If we decide to have APIs similar to JAX's pytree, I think the current implementation would be fine.

For performance, I would like to have two different path semantics, i.e., (1) path (tuple of raw indices); and (2) accessor (or typed path in #113378).

#### (1) path (tuple of raw indices)

Observation: Most nested containers are in the same type, e.g., nested `dict` or `torch.nn.Module`. The access rule to a container element is the same on every depth of the tree (`__getitem__` for `dict` and `__getattr__` for `torch.nn.Module`).

We can obtain the path entries during the tree flattening, which will only introduce a very small overhead (no auxiliary key path wrapper is applied). Because in most cases `context` and `entries` are the same.

```python
def dict_flatten(d):
    keys, values = list(d), list(d.value())
    return values, keys, keys  # children, context, entries
```

`optree` uses this approach to save the `entries` in the `treespec` object while keeping both `tree_flatten` and `tree_flatten_with_path` with good performance. This design also allows us to have only one `flatten_fn` definition in the registry. So we do not need to worry about the user-provided `flatten_fn` and `flatten_with_keys_fn` diverge.

```python
Children = Iterable[PyTree]
Context = Any
Entry = Any

flatten_fn(tree: PyTree) -> Union[
    Tuple[Children, Context],
    Tuple[Children, Context, Optional[Iterable[Entry]]]
]
```

See also: [OpTree: Registering a Container-like Custom Type as Non-leaf Nodes](https://github.com/metaopt/optree#registering-a-container-like-custom-type-as-non-leaf-nodes)

#### (2) accessor (tuple of path entries with semantics)

This is very similar to the API proposed in https://github.com/pytorch/pytorch/pull/116786#issuecomment-1888058691, I have an initial implementation for this https://github.com/XuehaiPan/optree/blob/c8002e7e984445589d7f34f12c9a03aa7f59622f/optree/registry.py#L565-L646. It is similar to [`itemgetter`](https://docs.python.org/3/library/operator.html#operator.itemgetter) and [`attrgetter`](https://docs.python.org/3/library/operator.html#operator.attrgetter) with `__call__` magic methods.

A typed path is `(entry, type, kind)`, it can be wrapped with `PyTreePathEntry`. A `PyTreeAccessor` is a tuple of `PyTreePathEntry`.

```python
class PyTreePathEntry:
    entry: Any
    type: Type
    kind: PyTreeKind

    def __call__(self, obj):
        # index into the object (based on kind or user-defined logic in a subclass)
        return obj[self.entry]


class PyTreeAccessor(tuple):  # Tuple[PyTreePathEntry, ...]
    def __call__(self, obj):
        for entry in self:
            obj = entry(obj)
        return obj
```

```python
In [1]: import optree

In [2]: from collections import *

In [3]: MyTuple = namedtuple('MyTuple', ['x', 'y', 'z'])

In [4]: tree = OrderedDict([('a', (1, 2)), ('b', {'c': [3, 4], 'd': 5}), ('e', MyTuple(6, 7, 8))])

In [5]: optree.tree_flatten_with_accessor(tree)
Out[5]: 
(
    [
        PyTreeAccessor(*['a'][0], (PyTreePathEntry(entry='a', type=<class 'collections.OrderedDict'>, kind=<PyTreeKind.ORDEREDDICT: 7>), PyTreePathEntry(entry=0, type=<class 'tuple'>, kind=<PyTreeKind.TUPLE: 3>))),
        PyTreeAccessor(*['a'][1], (PyTreePathEntry(entry='a', type=<class 'collections.OrderedDict'>, kind=<PyTreeKind.ORDEREDDICT: 7>), PyTreePathEntry(entry=1, type=<class 'tuple'>, kind=<PyTreeKind.TUPLE: 3>))),
        PyTreeAccessor(*['b']['c'][0], (PyTreePathEntry(entry='b', type=<class 'collections.OrderedDict'>, kind=<PyTreeKind.ORDEREDDICT: 7>), PyTreePathEntry(entry='c', type=<class 'dict'>, kind=<PyTreeKind.DICT: 5>), PyTreePathEntry(entry=0, type=<class 'list'>, kind=<PyTreeKind.LIST: 4>))),
        PyTreeAccessor(*['b']['c'][1], (PyTreePathEntry(entry='b', type=<class 'collections.OrderedDict'>, kind=<PyTreeKind.ORDEREDDICT: 7>), PyTreePathEntry(entry='c', type=<class 'dict'>, kind=<PyTreeKind.DICT: 5>), PyTreePathEntry(entry=1, type=<class 'list'>, kind=<PyTreeKind.LIST: 4>))),
        PyTreeAccessor(*['b']['d'], (PyTreePathEntry(entry='b', type=<class 'collections.OrderedDict'>, kind=<PyTreeKind.ORDEREDDICT: 7>), PyTreePathEntry(entry='d', type=<class 'dict'>, kind=<PyTreeKind.DICT: 5>))),
        PyTreeAccessor(*['e'][0], (PyTreePathEntry(entry='e', type=<class 'collections.OrderedDict'>, kind=<PyTreeKind.ORDEREDDICT: 7>), PyTreePathEntry(entry=0, type=<class '__main__.MyTuple'>, kind=<PyTreeKind.NAMEDTUPLE: 6>))),
        PyTreeAccessor(*['e'][1], (PyTreePathEntry(entry='e', type=<class 'collections.OrderedDict'>, kind=<PyTreeKind.ORDEREDDICT: 7>), PyTreePathEntry(entry=1, type=<class '__main__.MyTuple'>, kind=<PyTreeKind.NAMEDTUPLE: 6>))),
        PyTreeAccessor(*['e'][2], (PyTreePathEntry(entry='e', type=<class 'collections.OrderedDict'>, kind=<PyTreeKind.ORDEREDDICT: 7>), PyTreePathEntry(entry=2, type=<class '__main__.MyTuple'>, kind=<PyTreeKind.NAMEDTUPLE: 6>)))
    ],
    [1, 2, 3, 4, 5, 6, 7, 8],
    PyTreeSpec(OrderedDict([('a', (*, *)), ('b', {'c': [*, *], 'd': *}), ('e', MyTuple(x=*, y=*, z=*))]))
)

In [6]: accessors, leaves, treespec = optree.tree_flatten_with_accessor(tree)

In [7]: accessor = accessors[3]

In [8]: accessor
Out[8]: PyTreeAccessor(*['b']['c'][1], (PyTreePathEntry(entry='b', type=<class 'collections.OrderedDict'>, kind=<PyTreeKind.ORDEREDDICT: 7>), PyTreePathEntry(entry='c', type=<class 'dict'>, kind=<PyTreeKind.DICT: 5>), PyTreePathEntry(entry=1, type=<class 'list'>, kind=<PyTreeKind.LIST: 4>)))

In [9]: accessor(tree)
Out[9]: 4

In [10]: accessor.pprint('tree')
Out[10]: "tree['b']['c'][1]"

In [11]: tree['b']['c'][1]
Out[11]: 4

In [12]: accessor[:-1](tree)
Out[12]: [3, 4]

In [13]: accessor[:-2](tree)
Out[13]: {'c': [3, 4], 'd': 5}

In [14]: accessor[0](tree)
Out[14]: {'c': [3, 4], 'd': 5}

In [15]: tree['b']
Out[15]: {'c': [3, 4], 'd': 5}
```

zou3519 commented on test/test_pytree.py:
```diff
@@ -1056,6 +1060,127 @@ def test_saved_serialized(self):
+            actual = py_pytree.tree_unflatten([leaf for _, leaf in key_leaves], spec)
+            self.assertEqual(actual, pytree)
+
+    def test_key_str(self):
```
Can you add a test for tree_leaves_with_path?

zou3519 reviewed (commented):
Looks good, the main comments were:
- we should make sure everyone's on board with this keypath solution (that is, we'll implement the same one for C++ pytree)
- the registration looks a bit different from JAX's
- we should try to fold the registration into NodeDef, because otherwise, we just keep on getting new global variables flying around.

suo commented on torch/utils/_pytree.py:
```diff
@@ -126,6 +144,16 @@ class _SerializeNodeDef(NamedTuple):
+    flatten_with_keys: FlattenWithKeysFunc
+
+
+SUPPORTED_KEY_PATH_NODES: Dict[Type[Any], _KeyPathNodeDef] = {}
```
sg sg

suo committed:
Update on "[pytree] add key path api"


This PR introduces a key path API to pytrees, drawing direct inspiration from JAX's [key path API](https://jax.readthedocs.io/en/latest/jax-101/05.1-pytrees.html#key-paths).

I added the 3 APIs described there, and a registry of `flatten_with_keys` fns for each node type, which is a version of `flatten` that also returns `KeyEntry`s describing how to access values from the original pytree.

Current use cases for this API:
- Folks would like to do argument traversal over input pytrees to do verification and compatibility enforcement. Keypaths are useful for this—https://fburl.com/code/06p7zrvr is a handrolled pass doing basically the same thing but probably more fragilely.
- In export non-strict mode, we need to figure out a way to track sources for pytree inputs. In strict mode, dynamo handles this for us, but we'd like a decoupled component to handle this when we're not using dynamo.

I'm sure there are places it would be useful.

Some design notes:
- I only implemented the API for  the Python pytree impl. optree has some differences in how their keypath APIs are designed (see https://github.com/pytorch/pytorch/issues/113378 for discussion). I have some issues with the proposed typed_path solution in that discussion and prefer JAX's API, but we can hash that out separately.
- The way folks register a `flatten_with_keys` fn is through a new kwarg to `register_pytree_node`. This follows how we do serialization fns, although the list of additional arguments is getting unwieldy.
- My impl handles pytrees with an undefined `flatten_with_keys` fn is different from JAX. I will raise an error, JAX creates a fallback keyentry.

Differential Revision: [D52547850](https://our.internmc.facebook.com/intern/diff/D52547850/)

[ghstack-poisoned]

suo committed:
Update on "[pytree] add key path api"


This PR introduces a key path API to pytrees, drawing direct inspiration from JAX's [key path API](https://jax.readthedocs.io/en/latest/jax-101/05.1-pytrees.html#key-paths).

I added the 3 APIs described there, and a registry of `flatten_with_keys` fns for each node type, which is a version of `flatten` that also returns `KeyEntry`s describing how to access values from the original pytree.

Current use cases for this API:
- Folks would like to do argument traversal over input pytrees to do verification and compatibility enforcement. Keypaths are useful for this—https://fburl.com/code/06p7zrvr is a handrolled pass doing basically the same thing but probably more fragilely.
- In export non-strict mode, we need to figure out a way to track sources for pytree inputs. In strict mode, dynamo handles this for us, but we'd like a decoupled component to handle this when we're not using dynamo.

I'm sure there are places it would be useful.

Some design notes:
- I only implemented the API for  the Python pytree impl. optree has some differences in how their keypath APIs are designed (see https://github.com/pytorch/pytorch/issues/113378 for discussion). I have some issues with the proposed typed_path solution in that discussion and prefer JAX's API, but we can hash that out separately.
- The way folks register a `flatten_with_keys` fn is through a new kwarg to `register_pytree_node`. This follows how we do serialization fns, although the list of additional arguments is getting unwieldy.
- My impl handles pytrees with an undefined `flatten_with_keys` fn is different from JAX. I will raise an error, JAX creates a fallback keyentry.

Differential Revision: [D52547850](https://our.internmc.facebook.com/intern/diff/D52547850/)

[ghstack-poisoned]

suo committed:
Update on "[pytree] add key path api"


This PR introduces a key path API to pytrees, drawing direct inspiration from JAX's [key path API](https://jax.readthedocs.io/en/latest/jax-101/05.1-pytrees.html#key-paths).

I added the 3 APIs described there, and a registry of `flatten_with_keys` fns for each node type, which is a version of `flatten` that also returns `KeyEntry`s describing how to access values from the original pytree.

Current use cases for this API:
- Folks would like to do argument traversal over input pytrees to do verification and compatibility enforcement. Keypaths are useful for this—https://fburl.com/code/06p7zrvr is a handrolled pass doing basically the same thing but probably more fragilely.
- In export non-strict mode, we need to figure out a way to track sources for pytree inputs. In strict mode, dynamo handles this for us, but we'd like a decoupled component to handle this when we're not using dynamo.

I'm sure there are places it would be useful.

Some design notes:
- I only implemented the API for  the Python pytree impl. optree has some differences in how their keypath APIs are designed (see https://github.com/pytorch/pytorch/issues/113378 for discussion). I have some issues with the proposed typed_path solution in that discussion and prefer JAX's API, but we can hash that out separately.
- The way folks register a `flatten_with_keys` fn is through a new kwarg to `register_pytree_node`. This follows how we do serialization fns, although the list of additional arguments is getting unwieldy.
- My impl handles pytrees with an undefined `flatten_with_keys` fn is different from JAX. I will raise an error, JAX creates a fallback keyentry.

Differential Revision: [D52547850](https://our.internmc.facebook.com/intern/diff/D52547850/)

[ghstack-poisoned]

suo committed:
Update on "[pytree] add key path api"


This PR introduces a key path API to pytrees, drawing direct inspiration from JAX's [key path API](https://jax.readthedocs.io/en/latest/jax-101/05.1-pytrees.html#key-paths).

I added the 3 APIs described there, and a registry of `flatten_with_keys` fns for each node type, which is a version of `flatten` that also returns `KeyEntry`s describing how to access values from the original pytree.

Current use cases for this API:
- Folks would like to do argument traversal over input pytrees to do verification and compatibility enforcement. Keypaths are useful for this—https://fburl.com/code/06p7zrvr is a handrolled pass doing basically the same thing but probably more fragilely.
- In export non-strict mode, we need to figure out a way to track sources for pytree inputs. In strict mode, dynamo handles this for us, but we'd like a decoupled component to handle this when we're not using dynamo.

I'm sure there are places it would be useful.

Some design notes:
- I only implemented the API for  the Python pytree impl. optree has some differences in how their keypath APIs are designed (see https://github.com/pytorch/pytorch/issues/113378 for discussion). I have some issues with the proposed typed_path solution in that discussion and prefer JAX's API, but we can hash that out separately.
- The way folks register a `flatten_with_keys` fn is through a new kwarg to `register_pytree_node`. This follows how we do serialization fns, although the list of additional arguments is getting unwieldy.
- My impl handles pytrees with an undefined `flatten_with_keys` fn is different from JAX. I will raise an error, JAX creates a fallback keyentry.

Differential Revision: [D52547850](https://our.internmc.facebook.com/intern/diff/D52547850/)

[ghstack-poisoned]

suo committed:
Update on "[pytree] add key path api"


This PR introduces a key path API to pytrees, drawing direct inspiration from JAX's [key path API](https://jax.readthedocs.io/en/latest/jax-101/05.1-pytrees.html#key-paths).

I added the 3 APIs described there, and a registry of `flatten_with_keys` fns for each node type, which is a version of `flatten` that also returns `KeyEntry`s describing how to access values from the original pytree.

Current use cases for this API:
- Folks would like to do argument traversal over input pytrees to do verification and compatibility enforcement. Keypaths are useful for this—https://fburl.com/code/06p7zrvr is a handrolled pass doing basically the same thing but probably more fragilely.
- In export non-strict mode, we need to figure out a way to track sources for pytree inputs. In strict mode, dynamo handles this for us, but we'd like a decoupled component to handle this when we're not using dynamo.

I'm sure there are places it would be useful.

Some design notes:
- I only implemented the API for  the Python pytree impl. optree has some differences in how their keypath APIs are designed (see https://github.com/pytorch/pytorch/issues/113378 for discussion). I have some issues with the proposed typed_path solution in that discussion and prefer JAX's API, but we can hash that out separately.
- The way folks register a `flatten_with_keys` fn is through a new kwarg to `register_pytree_node`. This follows how we do serialization fns, although the list of additional arguments is getting unwieldy.
- My impl handles pytrees with an undefined `flatten_with_keys` fn is different from JAX. I will raise an error, JAX creates a fallback keyentry.

Differential Revision: [D52547850](https://our.internmc.facebook.com/intern/diff/D52547850/)

[ghstack-poisoned]

suo committed:
Update on "[pytree] add key path api"


This PR introduces a key path API to pytrees, drawing direct inspiration from JAX's [key path API](https://jax.readthedocs.io/en/latest/jax-101/05.1-pytrees.html#key-paths).

I added the 3 APIs described there, and a registry of `flatten_with_keys` fns for each node type, which is a version of `flatten` that also returns `KeyEntry`s describing how to access values from the original pytree.

Current use cases for this API:
- Folks would like to do argument traversal over input pytrees to do verification and compatibility enforcement. Keypaths are useful for this—https://fburl.com/code/06p7zrvr is a handrolled pass doing basically the same thing but probably more fragilely.
- In export non-strict mode, we need to figure out a way to track sources for pytree inputs. In strict mode, dynamo handles this for us, but we'd like a decoupled component to handle this when we're not using dynamo.

I'm sure there are places it would be useful.

Some design notes:
- I only implemented the API for  the Python pytree impl. optree has some differences in how their keypath APIs are designed (see https://github.com/pytorch/pytorch/issues/113378 for discussion). I have some issues with the proposed typed_path solution in that discussion and prefer JAX's API, but we can hash that out separately.
- The way folks register a `flatten_with_keys` fn is through a new kwarg to `register_pytree_node`. This follows how we do serialization fns, although the list of additional arguments is getting unwieldy.
- My impl handles pytrees with an undefined `flatten_with_keys` fn is different from JAX. I will raise an error, JAX creates a fallback keyentry.

Differential Revision: [D52547850](https://our.internmc.facebook.com/intern/diff/D52547850/)

[ghstack-poisoned]
